### ISSUE TITLE
Identifiers starting with underscores don't immediately tab-complete

### ISSUE THREAD

**Issue by jlstevens** [ioam/holoviews#36]
With our new identifier sanitization system, you can set labels that start with underscores or with special characters not normally allowed in attribute names. The result is a string that starts with a single underscore (never two though!) for use in the tree object (e.g. ``AttrTree``, ``Layout``, `Overlay``  or ``OptionTree``).

All this is fairly general but I am worried that this behavior may end up being confusing to users. The issue with attributes starting with a single underscore aren't presented for tab-completion in the notebook - at least not until you know that you have to suggest the initial underscore yourself.

This means that it may appear to users that elements on their Layouts have vanished even though they are actually available (but hidden). I have no way of changing the IPython tab-completion behavior so  one suggestion may be to generate a warning to discourage labels/group strings starting with underscores or special characters.

There could be a way to disable this warning but I'm not sure this is necessarily a good idea - the lack of obvious tab-completion when there is a leading underscore is present could be pretty confusing. It is better (in my opinion) to always warn people who are setting unsuitable labels/groups that they are doing something that is allowed but potentially very confusing to other people.

One other idea is that we could implement our own ``__dir__`` method which does give us some control over what IPython presents for tab-completion (still not underscore though). For instance, we could stop showing *all* the tab-completable attributes once a hidden one has been specified (the dictionary style interface is still available). At least this would help avoid tree where only some of the paths are easily visible while others are hidden.

Thoughts?

**Comment by jbednar** [comment]
I think that anyone who uses a label with an initial underscore can handle typing _ before tab complete; it's a pretty strange thing to do.  But having a warning in that case is also ok, for the same reason.  Would it work to have three underscores to represent the initial underscore?

**Comment by jlstevens** [comment]
> I think that anyone who uses a label with an initial underscore can handle typing _ before tab complete

Sure. When the user explicitly started with an underscore, it isn't too bad. The issue then is that any special character at the beginning of the label/group string are escaped such that the escaped string starts with an underscore that the user hadn't explicitly specified.

Does that suggest we should only warn when the attribute may be hidden without the user expecting it? In other words, warn if the string starts with a special character but not if the user explicitly specified an underscore at the start?

One way to alleviate the whole issue is to have some fixed (non-underscore) prefix to avoid generating an initial underscore that would be hidden. The problem is suggesting something sensible. I do have one quick idea though:

If  the string ``'$'`` is escaped to the hex code ``'_0x24_'`` maybe it could be escaped to 'hex24_' if at the start of the string? Sort of makes sense, avoids the initial underscore and doesn't make the escaped string any longer! I think this could be a good solution...

Not quite sure what you are suggesting by a triple underscore. Do you mean that for the user to specify a single leading underscore they would have to type three underscores instead? Not sure I follow!

**Comment by jbednar** [comment]
Ah, I see.  Yes, I was saying that they would type three underscores to match a single provided leading underscore, but that's not the important case, which is the underscores that we are creating and inserting ourselves for special characters.   Fundamentally, the problem is that Python identifiers must start with letters or underscores, and the convention is that a leading underscore denotes a less-suitable thing to tab-complete, so that just leaves letters, which of course are impossible to distinguish from user-supplied letter strings.  

It does sound like we have to generate a value that starts with a letter, to make tab completion work.  I don't much like the idea of using a different code for the initial character (hex vs. _0x for later characters), which seems tricky and harder to anticipate, but I guess that's an improvement.  Personally, I'd vote for just choosing some alphabetic prefix that we use for any string with a non-alphabetic initial character, but I can't think of any suitable such prefix...

**Comment by jlstevens** [comment]
How about truncating the ``_0`` bit at the start so it would start with ``x24_`` instead?

**Comment by jlstevens** [comment]
Ok, I've discussed this with Philipp and I think we've agreed on a solution:

* In Python 2, I will use the solution I propose immediately above ('x24' to start identifiers instead of the underscore).
* In Python 3, we can be far more relaxed as Python 3 does support a subset of Unicode as literals. For instance, ``f.α`` is perfectly valid syntax.

We still need to escape Python 3 identifiers as many unicode classes are not support as identifiers. Having different behavior between Python 2 and 3 is fine: although HoloViews itself supports both Python 2 and 3, there is never any guarantee that user code will be backwards/forwards compatible.

Fundamentally, we might as well make use of some of the improvements made in Python 3!

**Comment by philippjfr** [comment]
I agree that in principle being consistent about escaping special characters would be preferable. However if we decide on some custom scheme like ``x24_`` in general it's not immediately obvious that they are in fact escaped characters so I'm okay with mixing ``x24_`` for the first character and ``_0x24_`` elsewhere. We don't rely on unescaping these strings in any way since all they're used for is tab completion and the user can still supply the unescaped version and it will match it when using getitem and traverse.

So I think we should go with the solution above and make use of the unicode support in Python3.

**Comment by jbednar** [comment]
Sounds ok to me.

**Comment by jlstevens** [comment]
I've implemented the new system in these commits (oldest to newest): 2e955132 609a8454 1e8df6f9e f240f902e

In the end, after quite a bit of painful debugging, we found that we should definitely disable leading underscores in attribute names. Because IPython does a ``getattr`` looking for ``_ipython_display_`` on our objects, ``Layout`` responds by creating a node instead of raising `AttributeError`. Then IPython gets confused and then everything breaks (only on Python 3.4 for some reason). 

As identifiers starting with an underscore are hidden during tab-completion anyway and this is something that I don't thing many people will *want* it does seem like disabling leading underscores is a reasonable option (now implemented). For the same reason, character classes that are not normally usable at the start of a Python identifier are also disallowed (i.e digits in Python 2 and a few different unicode class in Python 3).

With the new code, the escaping behavior is pretty reasonable in Python 2 and only applied as necessary with Python 3 (you can use all your greek letters!). I think we have now given users as much flexibility in choosing their group/label strings as is reasonably possible in a simple system.

The only (optional!) extension I can think of is to allow customizable substitutions to be defined e.g something like

```python
substitutions = { '+':'plus'  , '-':'minus',  '/':'over',  '*':'times'}
```
I've not necessarily saying it is a *good* idea but it is certainly an idea. :-)

As these commits address the original issue I will close this issue shortly. If you want to discuss the idea of custom substitutions, that should be done in a new issue thread.

**Comment by jlstevens** [comment]
In the end, I've rewritten ``sanitize_identifier`` completely use a new approach that generates much nicer sanitized strings (2ffb8458).

I've add 42 new unit tests are here are some examples of the new behavior:

```python
>>> sanitize_identifier('Δ', version=2) # Python 2 version
'Delta'
>>> sanitize_identifier('Δ', version=3) # Python 3 version
'Δ'
>>> sanitize_identifier('δ', version=2) # Python 2 version
'delta'
>>> sanitize_identifier('δ', version=3) # Python 3 version
'δ'
>>> sanitize_identifier('α β', version=2)  # Python 2 version
'alpha_beta'
>>> sanitize_identifier('α β', version=3)  # Python 3 version
'α_β'
```

More examples are defined in the unit tests (ea555e6f). Hopefully, you'll agree that this is a big improvement!

**Comment by jbednar** [comment]
Looks good.  I wasn't able to follow the logic all the way through, but are there really things that are not allowed as labels?  Why would e.g. a leading underscore or a leading digit not just be sanitized away, so that they are fine as a label but would get translated just as sanitize does with other non-allowed characters?  Maybe I'm misinterpreting the flow of the logic, though.

**Comment by jlstevens** [comment]
Right. Some of the information can be found in this [reference](https://docs.python.org/3/reference/lexical_analysis.html#identifiers). There are two parts to this:

1. Symbols never allowed in Python identifiers that must always be sanitized. Python 3 allows more characters than Python 2 does.
2. Symbols allowed in identifiers but just not at the start of the identifier (i.e. digits in Python 2, categories Mn, Mc, Nd, Pc in Python 3).

Yes I could sanitize the latter case as well in order to allow strings starting with a leading digits (still want to disable underscores though) but I originally thought it would be ok to preserve *some* of the normal Python rules for identifiers. Honestly, I don't really mind either way. Which is better?

* Disallowing the string '88 cats' entirely (are strings starting with numbers that desirable?)
* Sanitizing '88 cats' which would become 'eight_8_cats' under the current approach.

I guess the fanciest approach would be to look for sequences of leading digits and then using a package such as [num2words](https://pypi.python.org/pypi/num2words) to make it into words in the appropriate language. Not saying we should do this though!

**Comment by jbednar** [comment]
I think you should allow any unicode string as a label, while sanitizing it as appropriate for use in attribute access.  Because labels are used by default as titles, it's not good if there are certain labels that work, and others that do not.  Leading digits are especially important -- people could very easily want to label their plots with a number, representing some condition or reference number, and we should certainly let them.  Leading underscores are less crucial, but they could come up as e.g. some Python variable or attribute name whose value is being plotted, and there's no point in users having to know or worry that such labels are not supported since we have this whole mechanism to sanitize labels anyway.  I strongly vote for simply sanitizing *everything* so that we can be done with it!

I don't think it's useful to convert a leading digit into the textual equivalent; having 8445567 turn into eight_445567 would be very strange.  Why not simply add a letter-based prefix if you detect a leading underscore or a leading digit, so that 8445567 turns into e.g. A_8445567 (where A_ is arbitrary, but is meant to stand for Attribute) and _i turns into A__i?  Seems trivial to implement, and much better behaved than trying to convert a digit into text or to have special rules for leading underscores.

**Comment by philippjfr** [comment]
We should have made you aware of some of the issues involved with allowing everything. Since AttrTree creates new leafs on getattr, it's quick to accumulate a lot of stuff. This is a particular problem when using them with IPython as the display and tab-completion mechanisms try to access a number of attributes. This means that we end up with a bunch of garbage underscore paths in the AttrTree.

As for numeric groups and labels, I really would like to discourage it. When you have numeric groups and labels that's a clear sign you should be using NdOverlay/NdLayout. If you use a number in your group or label that is still supported.

My vote is for disallowing both leading underscores due to the headaches with IPython it avoids and disabling groups/labels starting with numbers as that's what NdOverlay and NdLayout are for.

P.S.: There are no more hex character codes, unicodedata, Jean-Lucs implementation uses unicodedata.name to transcribe any non-valid identifier to it's unicode name, ``$`` becomes ``dollar``, ``*`` becomes ``asterisk`` etc..

**Comment by jlstevens** [comment]
I am ok with prefixing with ``A_`` for digits and just ``A`` for leading underscores. I suppose some system (even if totally arbitrary) is better that raising exceptions. 

I also agree with Philipp that numbers should often be done using NdLayouts or NdOverlays but I don't see much harm using the ``A`` or ``A_``prefix. These will be the first tab-completion options (first capital letter) and the prefix will only show up for leading underscores/digits and *not* all leading special characters (as would have been necessary in the old system).

**Comment by jbednar** [comment]
People can use numbers in ways that are not organized into a dimension, either because it's just the first part of a longer title, or if they represent e.g. animal numbers (which are numeric but unsorted).  So I don't think numbers == NdLayout.

**Comment by philippjfr** [comment]
Okay I don't have any larger objection to prefixing them with ``A_`` apart from it being weird. Might as well allow it though. I would probably still recommend that even Animal numbers are placed in an NdLayout, which can be sorted categorically in whatever is desired when supplying values to the Dimension. But I agree there is no harm in allowing it.

**Comment by jlstevens** [comment]
I'm only happy allowing this as this prefix as I expect it will only appear relatively rarely. It is still a strange thing to do so I am still tempted to prompt the user with a warning to choose something better.

**Comment by jlstevens** [comment]
Implemented in ed8ebdc2. You can check if the prefix will be used for a particular as follows:

```python
>>> sanitize_identifier.prefixed('_foo', version=2)
True
```

I think we can now close this issue now sanitization is now much improved and this new system also resolves the tab-completion issue. The only thing I think we need to do is add an FAQ item recommending against starting with underscore or digits.

**Comment by jbednar** [comment]
Much better, but I don't see why you added special handling for a leading underscore -- why not simply use a fixed prefix A_, such that _i becomes A__i (notice the two underscores)?  That uses less logic, gives the user fewer special cases to keep in their heads, and allows _i to be distinguished from i (as one might want to do if plotting Python variables, as I suggested before).  

Similarly, why are there still disallowed names?  Why not simply add the prefix to those as well?  

I strongly, strongly argue that we should simply accept whatever we are given, and make it be ok, using simple, predictable rules.  Please let the user just not have to worry about this!  We should be able to tell the user something straightforward, such as::

```
  Any legal Unicode string can be supplied as the label, and will be used to generate 
  an attribute name for use with attribute access.  Because only a subset of possible 
  strings are allowed as Python identifiers, the string will be sanitized to create a 
  valid identifier as appropriate for the Python version in use (i.e. label "α*3" will be 
  accessed as "alpha_times_3" under Python 2, but "α_times_3" under Python 3, 
  since Python 3 has a larger set of allowed characters).  Since Python does not 
  allow leading digits in identifiers, and to avoid problems with reserved words in 
  IPython, labels with leading digits, leading underscores, or that match certain 
  keywords will also be prefixed with A_ (e.g. "3a" will be "A_3a", and "_i" will be
  "A__i".  Tab-completion in IPython provides a convenient way to see how the 
  label has been translated into an attribute name.
```

If we could get this description any simpler, that would be even better, but this is definitely an improvement over the previous complex mix of rules, exceptions, and disallowed items.  Let's make the user's life simpler, since we are the users!

**Comment by jlstevens** [comment]
The two disallowed names must exist because everything breaks otherwise. I don't want to say more than that because the situation is too involved to bother explaining in full. Your suggestion for adding a prefix to those cases simply will not work (and can never work).

I don't really care about how the ``A`` prefix works because no one should ever supply strings that requires it. Just pick something else! If you really want the prefix to always be `A_` I can do that - makes no difference to me. Again, don't use it! Maybe the prefix should be ``'YOU_ARE_DOING_IT_WRONG_'``?

The text your wrote is ok but I don't think we need to say anything in the official tutorials. This is a side-note that goes somewhere but is not necessary on the main path for people learning HoloViews.

**Comment by jbednar** [comment]
It's very difficult for me to believe that if Trait_names is not allowed as an attribute, then A_Trait_names will also not be allowed; you'll have to convince me of that in our next meeting.  

In any case, I think it's extremely arrogant to say that people should simply pick something else for their label -- the labels are used for titles in displayed plots, and are not an internal detail like a variable name.  Yes, people can always change the title at the very end, just before plotting, but the whole point of HoloViews is to let people annotate their data with meaningful information, where *they*, not us, get to decide what is meaningful.  If people want to label their data with numbers, that's up to them!  Why on earth should we make it our problem?  It's not up to us to agree or disagree with how they label their data; that's beyond presumptuous!  We should  make recommendations, but they know much better than we do about what their data represents.  I know you're irritated that this is still an issue, but that does not make it somehow not be an issue.

And I don't think it's ok to hide this issue -- if we tell people that the labels are used for titles, and we tell them that the labels are used for attribute access, we have to also tell them that the attribute may not match the label.  It can be in the form of a link to some other text outside the tutorials (since that's not tutorial information), but it's definitely information that they need for them to get their work done.

**Comment by philippjfr** [comment]
Here is the problem, every single AttrTree would have a bunch of IPython attributes shoved on them unless we explicitly filter every single thing they decide to access on the objects now and in future. The tab-completion suggestions would look something like this (and with the A_ prefix they'd be right at the top too):

.A_repr_html
.A_repr_latex
.A_whatever_attributes_ipython_decides_to_getattr
...

And I don't think that's the whole problem. We've both tried removing the restriction a number of times and always ran into nothing but nightmares.

**Comment by jlstevens** [comment]
> In any case, I think it's extremely arrogant to say that people should simply pick something else for their label.

Recommending something that actually works isn't arrogant or presumptuous. Allowing people to pick names that will later either completely break all compositionality or result in strange/unexpected names is the terrible idea here.

Conventions exist for a reason and we aren't the ones who invented Python's conventions regarding what constitutes valid identifiers. If we introduce a convention, we do it for a reason: it is so that different bits code, written by different people works well together. Suggesting that we are at fault for trying to discourage bad ideas that constitute edge cases is simply preposterous. Sorry, but having Layouts that work is more important for us *and* other users than having some very *mild* conditions on the names you change (which are far more relaxed than what Python itself allows!).

Feel free to submit a pull request if you have a better idea.

**Comment by jbednar** [comment]
Philipp, it sounds to me like you're describing the opposite problem from the one I'm talking about.  The above seems to be about having .A__repr_html come into existence when someone accesses ._repr_html.  But I'm not proposing that such a thing.  A label is a parameter of a HoloViews object.  The attribute is something stored on the tree, and is by default constructed from the label.  The act of constructing the attribute is where A_ is added, isn't it?  So what does that have to do with attribute *access* done by IPython?  I'll ask that question of Jean-Luc in a minute.

Jean-Luc, what people title their plots is not edge cases or conventions or anything like that; it's simply completely up to them.  We don't get to decide how they want to label their plots.  We *do* get to decide on the attributes used to access those plots, which is completely up to us and what this code does.

**Comment by philippjfr** [comment]
The two are related problems due to the internals of AttrTree, as it uses getattr to create new paths. It would probably be possible to disentangle creation of new paths from regular getattrs but as changes in AttrTree always seem to cause a number of issues I'm not confident we can do that before this release.

**Comment by jlstevens** [comment]
@philippjfr : I think the only two ones we need to worry about are ``'Trait_names'`` and ``'_ipython_display_'``. I've never seen the other IPython related ``repr`` methods appear on any of our tree objects or cause any test failures.

**Comment by jbednar** [comment]
Ok, Jean-Luc and I have discussed it, and I think some things are clear:

1. Jean-Luc agreed to remove the special-case handling of an initial underscore, to make the logic simpler.  (I offered, but he said it would only take him a minute. :-)

2. At present, IPython checks for two special attributes, which are the disallowed ones for sanitize, and it does so in a way not compatible with assumptions made by AtrrTree.  Ideally, those two would be disallowed not at the sanitizing function, but at AttrTree itself, because they aren't specific to labels -- even a non-label-based method of setting such values as attributes would cause the same problem, as would defining any Python object with an attribute by those names.  So having them disallowed at the sanitize level is only a partial workaround, and is only to avoid having to edit  AttrTree.  This is not ideal, both because it's only a partial solution and because it needlessly complicates the user's idea about how attributes are computed from labels, but it's hopefully not a big deal because these two names are unlilkely to come up in practice.

3. We agreed that users should be encouraged to use labels that work well as Python identifiers, to make their lives easier.  Personally, I don't think we need any extra impediment such as warnings or scary names on top of this -- if they choose to have other kinds of labels, that's up to them; we just tell them what the consequences are.

So this issue should be done with, once J-L has tweaked the initial underscore bit.
